File: cpp/source/state/pilot/Disconnected.cpp
#include "state/pilot/Disconnected.hpp"

#include "event/input/user/Operate.hpp"

#include "state/pilot/Connecting.hpp"

using namespace fm;
using namespace fm::state;
using namespace fm::state::pilot;

using event::input::user::UserEvent;
using event::input::user::Operate;
using event::input::connection::ConnectionEvent;

using event::output::FacadeEvent;

Disconnected::Disconnected(IState& state) :
    IState(state)
{
}

Disconnected::Disconnected(IClient& client, IClient::Listener& listener, backend::ClientBackend& backend) :
    IState(client, listener, backend)
{
}

std::unique_ptr<IState> Disconnected::start()
{
    return nullptr;
}

std::unique_ptr<IState> Disconnected::handleUserEvent(const UserEvent& event)
{
    switch (event.getType())
    {
    case UserEvent::OPERATE:
    {
        const Operate& o = reinterpret_cast<const Operate&>(event);
        return std::make_unique<Connecting>(*this, o.getDeviceId(), o.getChannels());
    }

    default:
        return defaultEventHandle(event.toString());
    }
}

std::string Disconnected::toString() const
{
    return "Disconnected";
}

File: java/src/main/java/com/fleetmgr/sdk/client/state/pilot/Disconnected.java
package com.fleetmgr.sdk.client.state.pilot;

import com.fleetmgr.sdk.client.Client;
import com.fleetmgr.sdk.client.backend.ClientBackend;
import com.fleetmgr.sdk.client.event.input.connection.ConnectionEvent;
import com.fleetmgr.sdk.client.event.input.user.Operate;
import com.fleetmgr.sdk.client.event.input.user.UserEvent;
import com.fleetmgr.sdk.client.state.State;

/**
 * Created by: Bartosz Nawrot
 * Date: 23.09.2018
 * Description:
 */
public class Disconnected extends State {

    public Disconnected(Client client,
                        ClientBackend backend,
                        Client.Listener listener) {
        super(client, backend, listener);
    }

    Disconnected(State state) {
        super(state);
    }

    @Override
    public State start() {
        return null;
    }

    @Override
    public State notifyEvent(UserEvent event) {
        switch (event.getType()) {
            case OPERATE:
                Operate operate = (Operate)event;
                return new Connecting(this,
                        operate.getDeviceId(),
                        operate.getChannels());

            default:
                return defaultEventHandle(event.toString());
        }
    }

    @Override
    public State notifyConnection(ConnectionEvent event) {
        return defaultEventHandle(event.toString());
    }

    @Override
    public String toString() {
        return "Disconnected";
    }
}
